Keep track of which samples are filtered in the compendia

arielsvn opened the pull request:
## Issue Number

close #1838 

## Purpose/Implementation Notes

This adds `job_context['filtered_samples']` to keep track of the samples that are dropped at each step during the smashing pipeline. Then creates the file `filtered_samples_metadata.json` with the samples that were filtered and the reasons why - as described by https://github.com/alexslemonade/refinebio/issues/1838#issuecomment-559143625

## Types of changes

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Functional tests

Added tests for a compendium where we skip one sample.

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

kurtwheeler commented on workers/data_refinery_workers/processors/smashing_utils.py:
```diff
@@ -643,18 +667,22 @@ def write_non_data_files(job_context: Dict) -> Dict:
+            filtered_samples_path = os.path.join(job_context["output_dir"],
+                                                 'filtered_samples_metadata.json')
+            with open(filtered_samples_path, 'w',encoding='utf-8') as metadata_file:
+                json.dump(job_context['filtered_samples'], metadata_file, indent=4, sort_keys=True)
```
I'm not sure this metadata file is sufficient. For one thing, we tend to include a JSON version and a TSV version of every metadata file we include in the compendia, in case a user is only comfortable with one of the formats.

Secondly, in https://github.com/AlexsLemonade/refinebio/issues/1838#issuecomment-559143625 you mention

> We could also add all sample metadata there, and the experiment relations.

I would think we probably want to do this because if a user is actually using this file then they may be interested in analyzing why some samples got filtered. Would you agree with this @jaclyn-taroni? Should we include every metadata field we have about every sample that got filtered including it's experiment_accession_code?

arielsvn replied:
> we tend to include a JSON version and a TSV version of every metadata file we include in the compendia, in case a user is only comfortable with one of the formats.

We only have a tsv version for the samples inside the compendia. The main `aggregated_metadata.json` file doesn't have a tsv version.

![image](https://user-images.githubusercontent.com/1882507/70456498-37a44200-1a7c-11ea-90c1-4a6602566218.png)

I'd say the new `filtered_samples_metadata.json` file should be similar to `aggregated_metadata.json`. Do you think this makes sense?

> Should we include every metadata field we have about every sample that got filtered including it's experiment_accession_code?

Ok, I'll also include this information for each sample.

kurtwheeler commented on workers/data_refinery_workers/processors/smashing_utils.py:
```diff
@@ -304,6 +310,10 @@ def process_frames_for_key(key: str,
                                dataset_id=job_context['dataset'].id,
                                job_id=job_context["job"].id)
+                job_context['filtered_samples'][sample.accession_code] = {
+                    'reason': 'We were unable to smash the file associated with this sample during the first pass.',
```
This reason is going to go into the metadata file that we give our users. They won't know what "first pass" and "second pass" mean and we should avoid the use of the word "smash". A better reason to me would be: `The file associated with this sample did not pass the QC checks we apply before aggregating.` or something.

kurtwheeler commented on workers/data_refinery_workers/processors/smashing_utils.py:
```diff
@@ -74,6 +76,10 @@ def prepare_files(job_context: Dict) -> Dict:
                 found_files = True
+            else:
+                job_context['filtered_samples'][sample.accession_code] = {
+                    'reason': 'We could not find a valid smashable file for this sample.'
```
We shouldn't use `smashable` in copy destined for user consumption. Maybe `This sample did not have a processed file associated with it in our database.` would be better?

kurtwheeler commented on workers/data_refinery_workers/processors/smashing_utils.py:
```diff
@@ -375,9 +385,13 @@ def process_frames_for_key(key: str,
         if frame_data is None:
             job_context['unsmashable_files'].append(computed_file.filename)
+            job_context['filtered_samples'][sample.accession_code] = {
+                'reason': 'We were unable to smash the file associated with this sample during the second pass.',
```
Once again, they won't know what second pass means. Perhaps `The file associated with this sample did not contain a vector that fit the expected dimensions of the matrix.`? I'm less sure about this one. I honestly wouldn't think we'd see many of these.

arielsvn commented:
Thanks for reviewing @kurtwheeler. I think I addressed all your comments and included a new file named `filtered_samples_metadata.tsv`. Let me know if you see something else that can be improved

kurtwheeler commented:
These changes look good but looks like the tests aren't passing :(

arielsvn commented:
@kurtwheeler tests passing now :)
